### ISSUE TITLE
Crashes on startup - memlock

### ISSUE THREAD

**Issue by arkpope** [getumbrel/llama-gpt#7]
Hi, 

I am trying to run but I get this error after unning the 70B docker-compose.

System specs are Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz 12 Cores and 128Gb of RAM.

My best guess is that the docker file has to be modified to bypass mlock.

```:~/llama-gpt$ docker-compose -f docker-compose-70b.yml up
Starting llama-gpt_llama-gpt-api_1 ... done
Starting llama-gpt_llama-gpt-ui_1  ... done
Attaching to llama-gpt_llama-gpt-ui_1, llama-gpt_llama-gpt-api_1
llama-gpt-api_1  | python3 setup.py develop
llama-gpt-ui_1   | [INFO  wait] --------------------------------------------------------
llama-gpt-ui_1   | [INFO  wait]  docker-compose-wait 2.12.0
llama-gpt-ui_1   | [INFO  wait] ---------------------------
llama-gpt-ui_1   | [DEBUG wait] Starting with configuration:
llama-gpt-ui_1   | [DEBUG wait]  - Hosts to be waiting for: [llama-gpt-api:8000]
llama-gpt-ui_1   | [DEBUG wait]  - Paths to be waiting for: []
llama-gpt-ui_1   | [DEBUG wait]  - Timeout before failure: 600 seconds
llama-gpt-ui_1   | [DEBUG wait]  - TCP connection timeout before retry: 5 seconds
llama-gpt-ui_1   | [DEBUG wait]  - Sleeping time before checking for hosts/paths availability: 0 seconds
llama-gpt-ui_1   | [DEBUG wait]  - Sleeping time once all hosts/paths are available: 0 seconds
llama-gpt-ui_1   | [DEBUG wait]  - Sleeping time between retries: 1 seconds
llama-gpt-ui_1   | [DEBUG wait] --------------------------------------------------------
llama-gpt-ui_1   | [INFO  wait] Checking availability of host [llama-gpt-api:8000]
llama-gpt-ui_1   | [INFO  wait] Host [llama-gpt-api:8000] not yet available...
llama-gpt-api_1  | /usr/local/lib/python3.11/site-packages/setuptools/command/develop.py:40: EasyInstallDeprecationWarning: easy_install command is deprecated.
llama-gpt-api_1  | !!
llama-gpt-api_1  |
llama-gpt-api_1  |         ********************************************************************************
llama-gpt-api_1  |         Please avoid running ``setup.py`` and ``easy_install``.
llama-gpt-api_1  |         Instead, use pypa/build, pypa/installer or other
llama-gpt-api_1  |         standards-based tools.
llama-gpt-api_1  |
llama-gpt-api_1  |         See https://github.com/pypa/setuptools/issues/917 for details.
llama-gpt-api_1  |         ********************************************************************************
llama-gpt-api_1  |
llama-gpt-api_1  | !!
llama-gpt-api_1  |   easy_install.initialize_options(self)
llama-gpt-api_1  | [0/1] Install the project...
llama-gpt-api_1  | -- Install configuration: "Release"
llama-gpt-api_1  | -- Up-to-date: /app/_skbuild/linux-x86_64-3.11/cmake-install/llama_cpp/libllama.so
llama-gpt-api_1  | copying _skbuild/linux-x86_64-3.11/cmake-install/llama_cpp/libllama.so -> llama_cpp/libllama.so
llama-gpt-api_1  |
llama-gpt-api_1  | running develop
llama-gpt-api_1  | /usr/local/lib/python3.11/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
llama-gpt-api_1  | !!
llama-gpt-api_1  |
llama-gpt-api_1  |         ********************************************************************************
llama-gpt-api_1  |         Please avoid running ``setup.py`` directly.
llama-gpt-api_1  |         Instead, use pypa/build, pypa/installer or other
llama-gpt-api_1  |         standards-based tools.
llama-gpt-api_1  |
llama-gpt-api_1  |         See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
llama-gpt-api_1  |         ********************************************************************************
llama-gpt-api_1  |
llama-gpt-api_1  | !!
llama-gpt-api_1  |   self.initialize_options()
llama-gpt-api_1  | running egg_info
llama-gpt-api_1  | writing llama_cpp_python.egg-info/PKG-INFO
llama-gpt-api_1  | writing dependency_links to llama_cpp_python.egg-info/dependency_links.txt
llama-gpt-api_1  | writing requirements to llama_cpp_python.egg-info/requires.txt
llama-gpt-api_1  | writing top-level names to llama_cpp_python.egg-info/top_level.txt
llama-gpt-api_1  | reading manifest file 'llama_cpp_python.egg-info/SOURCES.txt'
llama-gpt-api_1  | adding license file 'LICENSE.md'
llama-gpt-api_1  | writing manifest file 'llama_cpp_python.egg-info/SOURCES.txt'
llama-gpt-api_1  | running build_ext
llama-gpt-api_1  | Creating /usr/local/lib/python3.11/site-packages/llama-cpp-python.egg-link (link to .)
llama-gpt-api_1  | llama-cpp-python 0.1.77 is already the active version in easy-install.pth
llama-gpt-api_1  |
llama-gpt-api_1  | Installed /app
llama-gpt-api_1  | Processing dependencies for llama-cpp-python==0.1.77
llama-gpt-api_1  | Searching for diskcache==5.6.1
llama-gpt-api_1  | Best match: diskcache 5.6.1
llama-gpt-api_1  | Processing diskcache-5.6.1-py3.11.egg
llama-gpt-api_1  | Adding diskcache 5.6.1 to easy-install.pth file
llama-gpt-api_1  |
llama-gpt-api_1  | Using /usr/local/lib/python3.11/site-packages/diskcache-5.6.1-py3.11.egg
llama-gpt-api_1  | Searching for numpy==1.25.1
llama-gpt-api_1  | Best match: numpy 1.25.1
llama-gpt-api_1  | Processing numpy-1.25.1-py3.11-linux-x86_64.egg
llama-gpt-api_1  | Adding numpy 1.25.1 to easy-install.pth file
llama-gpt-api_1  | Installing f2py script to /usr/local/bin
llama-gpt-api_1  | Installing f2py3 script to /usr/local/bin
llama-gpt-api_1  | Installing f2py3.11 script to /usr/local/bin
llama-gpt-api_1  |
llama-gpt-api_1  | Using /usr/local/lib/python3.11/site-packages/numpy-1.25.1-py3.11-linux-x86_64.egg
llama-gpt-api_1  | Searching for typing-extensions==4.7.1
llama-gpt-api_1  | Best match: typing-extensions 4.7.1
llama-gpt-api_1  | Adding typing-extensions 4.7.1 to easy-install.pth file
llama-gpt-api_1  |
llama-gpt-api_1  | Using /usr/local/lib/python3.11/site-packages
llama-gpt-api_1  | Finished processing dependencies for llama-cpp-python==0.1.77
llama-gpt-api_1  | Initializing server with:
llama-gpt-api_1  | Batch size: 2096
llama-gpt-api_1  | Number of CPU threads: 12
llama-gpt-api_1  | Number of GPU layers: 0
llama-gpt-api_1  | Context window: 4096
llama-gpt-ui_1   | [INFO  wait] Host [llama-gpt-api:8000] not yet available...
llama-gpt-api_1  | /usr/local/lib/python3.11/site-packages/pydantic/_internal/_fields.py:126: UserWarning: Field "model_alias" has conflict with protected namespace "model_".
llama-gpt-api_1  |
llama-gpt-api_1  | You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ('settings_',)`.
llama-gpt-api_1  |   warnings.warn(
llama-gpt-api_1  | llama.cpp: loading model from /models/llama-2-70b-chat.bin
llama-gpt-api_1  | llama_model_load_internal: format     = ggjt v3 (latest)
llama-gpt-api_1  | llama_model_load_internal: n_vocab    = 32000
llama-gpt-api_1  | llama_model_load_internal: n_ctx      = 4096
llama-gpt-api_1  | llama_model_load_internal: n_embd     = 8192
llama-gpt-api_1  | llama_model_load_internal: n_mult     = 4096
llama-gpt-api_1  | llama_model_load_internal: n_head     = 64
llama-gpt-api_1  | llama_model_load_internal: n_head_kv  = 64
llama-gpt-api_1  | llama_model_load_internal: n_layer    = 80
llama-gpt-api_1  | llama_model_load_internal: n_rot      = 128
llama-gpt-api_1  | llama_model_load_internal: n_gqa      = 1
llama-gpt-api_1  | llama_model_load_internal: rnorm_eps  = 1.0e-06
llama-gpt-api_1  | llama_model_load_internal: n_ff       = 24576
llama-gpt-api_1  | llama_model_load_internal: freq_base  = 10000.0
llama-gpt-api_1  | llama_model_load_internal: freq_scale = 1
llama-gpt-api_1  | llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama-gpt-api_1  | llama_model_load_internal: model size = 65B
llama-gpt-api_1  | llama_model_load_internal: ggml ctx size =    0.21 MB
llama-gpt-api_1  | warning: failed to mlock 221184-byte buffer (after previously locking 0 bytes): Cannot allocate memory
llama-gpt-api_1  | Try increasing RLIMIT_MLOCK ('ulimit -l' as root).
llama-gpt-api_1  | error loading model: llama.cpp: tensor 'layers.0.attention.wk.weight' has wrong shape; expected  8192 x  8192, got  8192 x  1024
llama-gpt-api_1  | llama_load_model_from_file: failed to load model
llama-gpt-api_1  | Traceback (most recent call last):
llama-gpt-api_1  |   File "<frozen runpy>", line 198, in _run_module_as_main
llama-gpt-api_1  |   File "<frozen runpy>", line 88, in _run_code
llama-gpt-api_1  |   File "/app/llama_cpp/server/__main__.py", line 46, in <module>
llama-gpt-api_1  |     app = create_app(settings=settings)
llama-gpt-api_1  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
llama-gpt-api_1  |   File "/app/llama_cpp/server/app.py", line 313, in create_app
llama-gpt-api_1  |     llama = llama_cpp.Llama(
llama-gpt-api_1  |             ^^^^^^^^^^^^^^^^
llama-gpt-api_1  |   File "/app/llama_cpp/llama.py", line 313, in __init__
llama-gpt-api_1  |     assert self.model is not None
llama-gpt-api_1  |            ^^^^^^^^^^^^^^^^^^^^^^
llama-gpt-api_1  | AssertionError
llama-gpt-api_1  | Exception ignored in: <function Llama.__del__ at 0x7f5f36b09e40>
llama-gpt-api_1  | Traceback (most recent call last):
llama-gpt-api_1  |   File "/app/llama_cpp/llama.py", line 1510, in __del__
llama-gpt-api_1  |     if self.ctx is not None:
llama-gpt-api_1  |        ^^^^^^^^
llama-gpt-api_1  | AttributeError: 'Llama' object has no attribute 'ctx'
llama-gpt_llama-gpt-api_1 exited with code 1
^CGracefully stopping... (press Ctrl+C again to force)
Stopping llama-gpt_llama-gpt-ui_1  ... done```

**Comment by notocwilson** [comment]
This is not a Dockerfile issue, but a container configuration issue. Since this container runs as root and won't lose context for ulimit adjustments, adjusting the docker-compose.yml file to set higher (or unlimited, though would not advise) ulimits, and potentially running the container at a higher privilege (depending on Docker version / system OS) may be in order.

**Comment by crackmac** [comment]
Try the answer in this issue.
https://github.com/getumbrel/llama-gpt/issues/4#issuecomment-1681330026

**Comment by arkpope** [comment]
I can confirm that the fix in https://github.com/getumbrel/llama-gpt/issues/4#issuecomment-1681330026H97342 solved the issue. Maybe we should add this?

**Comment by mayankchhabra** [comment]
Thanks for reporting, everyone! This was fixed by 2a24cb9e600ead49e2ef616d7f785eaf0613dc0b.
